### PR TITLE
transition cache leak during ledger catchup

### PR DESCRIPTION
**Pull request by ghost-not-in-the-shell**
1. Currently, if transitions received from peers didn't pass validation, those transitions are not getting invalidated in `unprocessed_transition_cache`.
2. In current implementation `ledger_catchup` would be blocked by transitions in cache, and the catchup would fail. This is super inefficient.
3. The subtree of catchup_scheduler that passed to ledger_catchup is not used effectively; besides instead of passing just 1 subtree, we should pass a forest.

This PR try to fix the above problems.

### CODE REVIEW

File: src/lib/transition_handler/validator.ml [hunk 1]
```diff
@@ -33,16 +35,14 @@ module Make (Inputs : Inputs.With_unprocessed_transition_cache.S) :
       |> External_transition.Verified.protocol_state
     in
     let%bind () =
-      Result.ok_if_true
-        (Transition_frontier.find frontier hash |> Option.is_none)
-        ~error:`Duplicate
+      Option.fold (Transition_frontier.find frontier hash) ~init:Result.ok_unit
```
Comment (johnwu93) [inline src/lib/transition_handler/validator.ml]: Why did you change this to `Option.fold`?

Comment (johnwu93) [reply src/lib/transition_handler/validator.ml]: Can you change the polymorphic variant `Duplicate to `In_frontier

File: src/lib/cache_lib/intf.ml [hunk 2]
```diff
@@ -1,6 +1,8 @@
 open Async_kernel
 open Core_kernel
 
+type 'a final_result = [`Failed | `Applied of 'a] Ivar.t
```
[the remark this answers is not in the document]
Comment (johnwu93) [reply src/lib/cache_lib/intf.ml]: I would call the polymorphic variant `Applied to be `Success

File: src/lib/cache_lib/impl.ml [hunk 3]
```diff
@@ -127,29 +120,21 @@ module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
...
     let invalidate (type a b) (t : (a, b) t) : a =
-      consume t ;
       mark_failed t ;
       Cache.remove (cache t) (original t) ;
       value t
 
     let free (type a b) (t : (a, b) t) : a =
-      consume t ;
-      mark_applied t ;
+      mark_success t ;
```
Comment (johnwu93) [inline src/lib/cache_lib/impl.ml]: Can we still call `assert_not_consumed t "cannot consume Cached.t twice" ;` when we try to invalidate cached item that has already been invalidated?

File: src/lib/ledger_catchup/ledger_catchup.ml [hunk 4]
```diff
@@ -186,23 +186,22 @@ module Make (Inputs : Inputs.S) :
...
           "transition queried during ledger catchup has already been seen" ;
         Deferred.return @@ Ok (Either.First breadcrumb)
-    | Error (`Under_processing final_result) -> (
+    | Error (`Under_processing consumed_state) -> (
         Logger.trace logger ~module_:__MODULE__ ~location:__LOC__
           "transition queried during ledger catchup is still under procession" ;
-        match%map Ivar.read final_result with
+        match%map Ivar.read consumed_state with
         | `Failed ->
+            (* TODO cancel the waiting job in both ledger_catch and \
```
Comment (johnwu93) [inline src/lib/ledger_catchup/ledger_catchup.ml]: `TODO:`

File: src/lib/transition_handler/catchup_scheduler.ml [hunk 5]
```diff
@@ -48,52 +51,37 @@ module Make (Inputs : Inputs.S) = struct
...
       ~catchup_breadcrumbs_writer =
     let collected_transitions = State_hash.Table.create () in
     let parent_root_timeouts = State_hash.Table.create () in
     let breadcrumb_builder_supervisor =
-      Capped_supervisor.create ~job_capacity:5 (fun transition_branches ->
-          build_breadcrumbs ~logger ~frontier transition_branches
-          >>= Writer.write catchup_breadcrumbs_writer )
+      Capped_supervisor.create ~job_capacity:5
+        (fun (initial_hash, transition_branches) ->
+          (* TODO: refact this to use Transition_handler.Breadcrumb_builder.build_subtrees_of_breadcrumbs
```
Comment (johnwu93) [inline src/lib/transition_handler/catchup_scheduler.ml]: Can you remove this TODO?

File: src/lib/transition_handler/breadcrumb_builder.ml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                        `Constructed new_breadcrumb
+                    | Error (`Fatal_error exn) -> Or_error.of_exn exn
+                    | Error (`Validation_error error) -> Error error )
+                |> Cached.sequence_deferred
+              in
+              Cached.sequence_result cached_result )
+        in
+        Rose_tree.map subtree_of_constructions ~f:(fun construction ->
+            Cached.transform construction ~f:(function
+              | `Initial -> failwith "impossible"
```
Comment (johnwu93) [inline src/lib/transition_handler/breadcrumb_builder.ml]: Is it possible to write this function without `Initial. It doesn't feel necessary and you are left with this failwith in the end.

File: src/lib/ledger_catchup/ledger_catchup.ml [hunk 7]
```diff
@@ -289,19 +182,25 @@ module Make (Inputs : Inputs.S) :
 
   let run ~logger ~network ~frontier ~catchup_job_reader
       ~catchup_breadcrumbs_writer ~unprocessed_transition_cache =
-    Strict_pipe.Reader.iter catchup_job_reader ~f:(fun subtree ->
-        match%bind
-          get_transitions_and_compute_breadcrumbs ~logger ~network ~frontier
-            ~num_peers:8 ~unprocessed_transition_cache ~target_subtree:subtree
-        with
-        | Ok tree -> Strict_pipe.Writer.write catchup_breadcrumbs_writer [tree]
+    Strict_pipe.Reader.iter catchup_job_reader ~f:(fun forest ->
```
Comment (johnwu93) [inline src/lib/ledger_catchup/ledger_catchup.ml]: Can you write this explicitly as a tuple? I thought this variable was just a forest, but it is a forest with transition

File: src/lib/transition_handler/validator.ml [hunk 8]
```diff
@@ -33,16 +33,14 @@ module Make (Inputs : Inputs.With_unprocessed_transition_cache.S) :
...
     let%bind () =
-      Result.ok_if_true
-        (not
-           (Unprocessed_transition_cache.mem unprocessed_transition_cache
-              transition_with_hash))
-        ~error:`Duplicate
+      Option.fold
+        (Unprocessed_transition_cache.final_state unprocessed_transition_cache
+           transition_with_hash) ~init:Result.ok_unit ~f:(fun _ final_state ->
+          Result.Error (`Under_processing final_state) )
```
Comment (johnwu93) [inline src/lib/transition_handler/validator.ml]: Can you call this variant, `In_process. Sounds grammatically more correct.

File: src/lib/ledger_catchup/ledger_catchup.ml [hunk 9]
```diff
@@ -184,49 +71,53 @@ module Make (Inputs : Inputs.S) :
...
+    match%bind cached_verified_transition with
+    | Ok x -> Deferred.return @@ Ok (Either.Second x)
+    | Error (`In_frontier hash) ->
+        Logger.trace logger ~module_:__MODULE__ ~location:__LOC__
           "transition queried during ledger catchup has already been seen" ;
-        Ok None
+        Deferred.return @@ Ok (Either.First hash)
+    | Error (`Under_processing consumed_state) -> (
+        Logger.trace logger ~module_:__MODULE__ ~location:__LOC__
+          "transition queried during ledger catchup is still under procession" ;
```
Comment (johnwu93) [inline src/lib/ledger_catchup/ledger_catchup.ml]: "transition queried during ledger catchup is still in process in one of the componenets in the transition_frontier"

File: src/lib/ledger_catchup/ledger_catchup.ml [hunk 10]
```diff
@@ -254,32 +146,33 @@ module Make (Inputs : Inputs.S) :
...
-                      Error (Error.of_string error) )
-                in
-                let full_subtree =
-                  List.fold_right verified_transitions ~init:target_subtree
-                    ~f:(fun transition acc -> Rose_tree.T (transition, [acc])
-                  )
+                let subtrees_of_transitions =
+                  if List.length verified_transitions > 0 then
+                    let rest, target_transition_singleton =
+                      List.split_n verified_transitions
```
Comment (johnwu93) [inline src/lib/ledger_catchup/ledger_catchup.ml]: The creation of `subtrees_of_transitions` is hard to read, namely because of `split_n`. Can you rewrite this without using `split_n` and `hd_exn`?

File: src/lib/transition_frontier_controller_tests/test_ledger_catchup.ml [hunk 11]
```diff
@@ -122,9 +124,116 @@ let%test_module "Ledger catchup" =
               ~f:Fn.id
             |> Option.value_exn
           in
-          let%map result =
-            test_catchup ~logger ~network me best_transition
-              (Rose_tree.of_list_exn @@ Non_empty_list.tail history)
+          test_catchup ~logger ~network me best_transition
+            (Rose_tree.of_list_exn @@ Non_empty_list.tail history) )
+
```
Comment (johnwu93) [inline src/lib/transition_frontier_controller_tests/test_ledger_catchup.ml]: please don't use initials to represent a variable.

File: src/config/test_postake_holy_grail.mlh [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+[%%define time_offsets true]
+[%%define fake_hash true]
+
+[%%define genesis_ledger "test_five_even_stakes"]
+[%%define genesis_state_timestamp "2019-01-30 12:00:00-08:00"]
+[%%define block_window_duration 1500]
+
+[%%define integration_tests true]
+[%%define force_updates false]
```
Comment (johnwu93) [inline src/config/test_postake_holy_grail.mlh]: add a newline here

File: src/lib/transition_frontier_controller_tests/test_ledger_catchup.ml [hunk 13]
```diff
@@ -122,9 +124,172 @@ let%test_module "Ledger catchup" =
...
+          let cached_failing_transition =
+            Transition_handler.Unprocessed_transition_cache.register
+              unprocessed_transition_cache failing_transition
+          in
+          Ledger_catchup.run ~logger ~network ~frontier:me
+            ~catchup_breadcrumbs_writer ~catchup_job_reader
+            ~unprocessed_transition_cache ;
+          let%bind _ = after (Core.Time.Span.of_ms 500.) in
+          Cache_lib.Cached.invalidate cached_failing_transition |> ignore ;
+          Deferred.return true )
```
Comment (johnwu93) [inline src/lib/transition_frontier_controller_tests/test_ledger_catchup.ml]: Shouldn't wait for all the other breadcrumbs in `history` to be invalidated?

**Comment by johnwu93** [comment]
Also, after I approve this PR, can you update the documentation for `Ledger_catchup`? It's slightly different after this PR update.

**Comment by nholland94** [comment]
I'm trying to understand the whole `final_state` stuff that was added to the cache stuff. It seems like the only way it's used is just in a single test? If that's the case, is adding this really necessary? This adds a pretty good deal of complexity to the abstraction, and some of the changes made me question whether or not the cache can leak now, although I need to look closer and understand it better. Just some initial questions upon starting reviewing.

File: src/lib/cache_lib/impl.ml [hunk 14]
```diff
@@ -95,33 +114,47 @@ module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
...
-      let%map () = Cache.remove (cache t) (original t) in
+    let invalidate (type a b) (t : (a, b) t) : a =
+      assert_not_finalized t "Cached item has already been finalized" ;
+      mark_failed t ;
+      Cache.remove (cache t) (original t) ;
+      value t
+
+    let free (type a b) (t : (a, b) t) : a =
+      assert_not_finalized t "Cached item has already been finalized" ;
+      mark_success t ;
```
Comment (nholland94) [inline src/lib/cache_lib/impl.ml]: I'm not sure if I find the names `invalidate` and `free` to be semantically accurate here. `invalidate` originally referred to "cache invalidation", which is just the removal of an item from the cache. Now it is being used to be imply a negative, as in there was a failure. Also, `free` seems to imply the release of a resource, which sort of makes sense, but doesn't imply a positive for the success. Perhaps you could pick some better names that directly imply some sort of negative and positive action?

File: src/lib/cache_lib/impl.ml [hunk 15]
```diff
@@ -13,49 +12,54 @@ module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
...
       let logger = Logger.extend logger [("cache", `String name)] in
       {name; set; logger}
 
+    let final_state t x = Hashtbl.find t.set x
+
     let register t x =
-      let%map () = Hash_set.strict_add t.set x in
-      Cached.create t x
+      let final_state = Ivar.create () in
+      Hashtbl.add_exn t.set ~key:x ~data:final_state ;
```
Comment (nholland94) [inline src/lib/cache_lib/impl.ml]: It would be nice to maintain the old interface of returning an `Or_error.t` instead of throwing exceptions.

File: src/lib/transition_handler/validator.ml [hunk 16]
```diff
@@ -33,16 +33,14 @@ module Make (Inputs : Inputs.With_unprocessed_transition_cache.S) :
...
+      Option.fold (Transition_frontier.find frontier hash) ~init:Result.ok_unit
+        ~f:(fun _ _ -> Result.Error (`In_frontier hash) )
     in
     let%bind () =
-      Result.ok_if_true
-        (not
-           (Unprocessed_transition_cache.mem unprocessed_transition_cache
-              transition_with_hash))
-        ~error:`Duplicate
+      Option.fold
```
Comment (nholland94) [inline src/lib/transition_handler/validator.ml]: I think there is a more canonical/readable alternative to using `Option.fold` like this (here and above):

```ocaml
Result.ok_if_true (Option.is_none your_option) ~error:your_error
```

For example, the above usage would become:

```ocaml
Result.ok_if_true (Option.is_none (Transition_frontier.find frontier hash)) ~error:(`In_frontier hash)
```
